### ISSUE TITLE
CPU High Load when Many PPP Interfaces are UP

### ISSUE THREAD

**Issue by Andlui9** [net-snmp/net-snmp#17]
Well, I've been using linux servers for a long time as a pppoe concentrator, and for some time I want to be able to monitor server resources, however, when I try to use the net-snmp daemon (on Debian I install it via snmpd package), the CPU consumption instantly climbs to the top in one of the CPU colors and is impossible to use. Does anyone know if anything can be done to fix it? Maybe some way to disable the collection of certain interfaces? I read somewhere that this has to do with "interface caching", if that's the same, what, can you disable?

**Comment by stannous** [comment]
Hi André,

I have a potential fix for this.  That is to introduce a new config option
"include_ifmib_iface_prefix" so that you can limit which interfaces
the IF-MIB processing will handle (e.g. "include_ifmib-iface_prefix eth lo wlan"
should ignore all your ppp interfaces and only track counters for these ifnames).

Please see my patch at 

https://github.com/net-snmp/net-snmp/pull/24

Regards,
Sam Tannous

**Comment by Andlui9** [comment]
I did not quite understand how to use conf, just add in the configuration file at any point? like this:

include_ifmib-iface_prefix eth ens enp dummy bond

PS: I'm using debian, and I applied the patch directly to the source code of the community maintained package and compiled it with debian mode.

**Comment by stannous** [comment]
Correct.  The patch, when applied to a recent (around 5.8.1) source tree and recompiled, will give you the new configuration option "include_ifmib_iface_prefix" (note the use of underscores and not a dash "-").  You would add this to your snmpd.conf file and restart snmpd.

This is currently just a pull-request so some things may change before the patch is accepted.

**Comment by stannous** [comment]
There is also the new option ifmib_max_num_ifaces NUM 
I just added.  See this patch:

```
  commit 6624b240e034e8f8a85acd09d19519150df2c0a3
        Author: Sam Tannous <stannous@cumulusnetworks.com>
        Date:   Wed Nov 13 11:20:46 2019 -0800
    
            This patch introduces a new config option "ifmib_max_num_ifaces NUM"
            This patch sets the maximum number of interfaces included in IF-MIB
            data collection.  For servers with a large number of interfaces
            (ppp, dummy, bridge, etc) the IF-MIB processing will take a large
            chunk of CPU for ioctl calls (on Linux). Setting a reasonable
            maximum number of ifaces for the CPU used will reduce the CPU
            load for IF-MIB processing.  For example, configuring
            "ifmib_max_num_ifaces 500" will include only the first 500
            interfaces based on ifindex and ignore all others for IF-MIB
            processing.
    
            Signed-off-by: Sam Tannous <stannous@cumulusnetworks.com>
            [ bvanassche: combined two patches into one and removed one unused variable ]
```

**Comment by Andlui9** [comment]
> There is also the new option ifmib_max_num_ifaces NUM
> I just added. See this patch:
> 
> ```
>   commit 6624b240e034e8f8a85acd09d19519150df2c0a3
>         Author: Sam Tannous <stannous@cumulusnetworks.com>
>         Date:   Wed Nov 13 11:20:46 2019 -0800
>     
>             This patch introduces a new config option "ifmib_max_num_ifaces NUM"
>             This patch sets the maximum number of interfaces included in IF-MIB
>             data collection.  For servers with a large number of interfaces
>             (ppp, dummy, bridge, etc) the IF-MIB processing will take a large
>             chunk of CPU for ioctl calls (on Linux). Setting a reasonable
>             maximum number of ifaces for the CPU used will reduce the CPU
>             load for IF-MIB processing.  For example, configuring
>             "ifmib_max_num_ifaces 500" will include only the first 500
>             interfaces based on ifindex and ignore all others for IF-MIB
>             processing.
>     
>             Signed-off-by: Sam Tannous <stannous@cumulusnetworks.com>
>             [ bvanassche: combined two patches into one and removed one unused variable ]
> ```

Hi!

I got the code here, but I didn't find this commit, can you help me?

**Comment by bvanassche** [comment]
Does this help?
~~~
$ git branch -r --contains 6624b240e034e8f8a85acd09d19519150df2c0a3 | grep github-official
  github-official/V5-8-patches
  github-official/master
~~~
